src/Orleans.Journaling/StateMachineManager.cs

```diff
@@ -0,0 +1,457 @@
+                                {
+                                    await _storage.ReplaceAsync(logSegment, cancellationToken).ConfigureAwait(false);
+                                }
+                                else
+                                {
+                                    await _storage.AppendAsync(logSegment, cancellationToken).ConfigureAwait(false);
+                                }
+
+                                // Notify all state machines that the operation completed.
+                                lock (_lock)
+                                {
+                                    foreach (var stateMachine in _stateMachines.Values)
+                                    {
+                                        stateMachine.OnWriteCompleted();
+                                    }
+                                }
+                            }
+                        }
+                        else if (workItem.Type is WorkItemType.DeleteState)
+                        {
+                            // Clear storage.
+                            await _storage.DeleteAsync(cancellationToken).ConfigureAwait(false);
+
+                            lock (_lock)
+                            {
+                                // Reset the state machine id collection.
+                                _stateMachineIds.ResetVolatileState();
+
+                                // Allocate new state machine ids for each state machine.
+                                // Doing so will trigger a reset, since _stateMachineIds will call OnSetStateMachineId, which resets the state machine in question.
+                                _nextStateMachineId = 1;
+                                foreach (var (name, stateMachine) in _stateMachines)
+                                {
+                                    var id = _nextStateMachineId++;
+                                    _stateMachineIds[name] = id;
+                                }
+                            }
+                        }
+                        else if (workItem.Type is WorkItemType.Initialize)
+                        {
+                            lock (_lock)
+                            {
+                                _state = ManagerState.Ready;
+                            }
+                        }
+                        else if (workItem.Type is WorkItemType.RegisterStateMachine)
+                        {
+                            lock (_lock)
+                            {
+                                if (_state is not ManagerState.Unknown)
+                                {
+                                    throw new InvalidOperationException("Registering a state machine after activation is invalid");
+                                    /*
+                                    // Re-enqueue the work item without completing it, while waiting for the state machine manager to be initialized.
```

Comment:
Should we re-enqueue instead of throwing? With some mitigation for infinite looping...